lib/services/local/session.go

```diff
@@ -95,3 +97,249 @@ func (s *IdentityService) DeleteAllAppSessions(ctx context.Context) error {
+	return s.WebSessions().Get(context.TODO(), types.GetWebSessionRequest{
+		User:      user,
+		SessionID: sessionID,
+	})
+}
+
+// UpsertWebSession updates or inserts a web session for the user and the session ID
+func (s *IdentityService) UpsertWebSession(user, sessionID string, session services.WebSession) error {
+	session.SetUser(user)
+	session.SetName(sessionID)
+	return s.WebSessions().Upsert(context.TODO(), session)
+}
+
+// DeleteWebSession deletes web session from the storage.
+func (s *IdentityService) DeleteWebSession(user, sessionID string) error {
+	return trace.Wrap(s.WebSessions().Delete(context.TODO(), types.DeleteWebSessionRequest{
+		User:      user,
+		SessionID: sessionID,
+	}))
+}
+
+// WebSessions returns the web sessions manager.
+func (s *IdentityService) WebSessions() services.WebSessionInterface {
+	return &webSessions{backend: s.Backend, log: s.log}
+}
+
+// Get returns the web session state described with req.
+func (r *webSessions) Get(ctx context.Context, req types.GetWebSessionRequest) (services.WebSession, error) {
+	if err := req.Check(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	item, err := r.backend.Get(ctx, webSessionKey(req.SessionID))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	session, err := services.GetWebSessionMarshaler().UnmarshalWebSession(item.Value, services.SkipValidation())
+	if err != nil && !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+	if session != nil {
+		return session, nil
+	}
+	// DELETE IN 7.x:
+	// Return web sessions from a legacy path under /web/users/<user>/sessions/<id>
+	return getLegacyWebSession(ctx, r.backend, req.User, req.SessionID)
+}
+
+// List gets all regular web sessions.
+func (r *webSessions) List(ctx context.Context) (out []services.WebSession, err error) {
+	key := backend.Key(webPrefix, sessionsPrefix)
+	result, err := r.backend.GetRange(ctx, key, backend.RangeEnd(key), backend.NoLimit)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	for _, item := range result.Items {
+		session, err := services.GetWebSessionMarshaler().UnmarshalWebSession(item.Value, services.SkipValidation())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		out = append(out, session)
+	}
+	// DELETE IN 7.x:
+	// Return web sessions from a legacy path under /web/users/<user>/sessions/<id>
+	legacySessions, err := r.listLegacySessions(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return append(out, legacySessions...), nil
+}
+
+// Upsert updates the existing or inserts a new web session.
+func (r *webSessions) Upsert(ctx context.Context, session services.WebSession) error {
+	value, err := services.GetWebSessionMarshaler().MarshalWebSession(session)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	sessionMetadata := session.GetMetadata()
+	item := backend.Item{
+		Key:     webSessionKey(session.GetName()),
+		Value:   value,
+		Expires: backend.EarliestExpiry(session.GetBearerTokenExpiryTime(), sessionMetadata.Expiry()),
```

Comment:
Nit: Doesn't resource already have `GetExpiryTime()` method, so you prob. don't need to extract metadata first?

Reply:
I kept the original behavior as much as possible where the `Metadata.Expires` is used as a resource expiration timestamp both ways. Although in case of web sessions the Spec.Expires is sort of redundant b/c of that. I'm on the fence but would probably be keen on keeping the redundancy.